Reconciler docu (#1083)
kyma-incubator/reconciler: README.md

|Scheduler|Queries the cluster inventory for clusters which require a reconciliation and adds them to the reconciliation queue.|
|Reconciliation Repository|Stores reconciliation related data, offers CRUD operations and manages the reconciliation queue.|
|Worker Pool|Queries the reconciliation repository for processable operations|
|Worker|Runs an reconciliation operation by creating and executing an invoker|
|Invoker|Calls the component reconciler (either remotely via an REST call or, if running embedded, locally by triggering a component reconciler instance)|
|Bookkeeper|Monitors the progress of cluster reconciliations (scans for completed cluster reconciliations, detects orphan operations etc.)|
|Cleaner|Removes outdated entries from the cluster inventory, shortens change history of clusters etc.|
|Metrics Collector|Tracks monitoring metrics|
|Occupancy Repository|Stores the occupancy of component reconciler worker pools|
|Metrics Exporter|Exports monitoring metrics as [Prometheus](https://prometheus.io/) scrape endpoint|

### 5.2.2 Component reconciler

|Component|Description|
|---|---|
|Worker Pool|Processes an reconciliation operation|

> |Runner|Executes the reconciliation of an operation|

Reviewer: ```suggestion
|Runner|Executes the reconciliation of a reconciliation operation.|
```

> |Chart Provider|Renders the HELM chart of an component|

Reviewer: ```suggestion
|Chart Provider|Renders the HELM chart of a component.|
```

> |Workspace Factory|Makes HELM charts locally available by downloading them for remote locations (e.g. cloud storages, Git repositories etc.)|

Reviewer: ```suggestion
|Workspace Factory|Makes HELM charts locally available by downloading them for remote locations (for example, cloud storages, Git repositories etc.)|
```

> |Heartbeat|Trigger the callback to send a reconciliation status of an operation to the mothership reconciler|

Reviewer: ```suggestion
|Heartbeat|Triggers the callback to send a reconciliation status of an operation to the mothership reconciler.|
```

> |Callback|Report the reconciliation status to the mothership (either via REST call or, if running embedded, by exectuing a callback function)|

Reviewer: ```suggestion
|Callback|Reports the reconciliation status to the mothership (either via REST call or, if running embedded, by executing a callback function).|
```

> |Metrics Exporter|Exports monitoring metrics as [Prometheus](https://prometheus.io/) scrape endpoint|

Reviewer: ```suggestion
|Metrics Exporter|Exports monitoring metrics as [Prometheus](https://prometheus.io/) scrape endpoint.|
```

|SRE / On-Call Engineer|Developer or SRE engineer who administrates the KCP microservices.|

# 8. Crosscutting Concepts

## 8.1 Logging

## 8.2 Security

The interaction between microservices in KCP is only allowed over secured channels (TLS encrypted connections).

Authentication is also mandatory for any communication and ensure either by

* mTLS (primarily used for Kubernetes internal communication between services via the Istio service mesh)
* SSL Key authentication (used when communicating to the Kubernetes cluster)
* protocol specific authentication used by the Postgres DB
* OAuth2 (used by the KCP CLI to authenticate developers or SRE engineers)

## 8.3 Monitoring

# 9. Architecture Decisions

## 9.1 Relational database as central storage

## 9.2 Tasks triggered in intervals

## 9.3 Pre-requisite components

## 9.4 Software layers

# 10. Risks and Technical Debt

## 10.1 Database

### 10.1.1 Custom ORM

### 10.1.2 Database as performance bottleneck

# 11. Glossary

|Acronym|Full name|Description|
|---|---|---|
|SKR|SAP Kyma Runtime|Managed Kyma cluster offering by SAP|
|KCP|Kymca Control Plane|Managing the lifecycle of SKRs|

> |KEB|Kyma Environment Broker|Service Broker implementation within BPT which manages business process related to SKRs.|

Reviewer: ```suggestion
|KEB|Kyma Environment Broker|Service Broker implementation within BPT, which manages business process related to SKRs.|
```

## 8.1 Logging

> The reconciler is following the recommended approach for [consistent logging in Kyma](https://github.com/kyma-project/community/tree/main/concepts/observability-consistent-logging).

Reviewer: ```suggestion
The reconciler follows the recommended approach for [consistent logging in Kyma](https://github.com/kyma-project/community/tree/main/concepts/observability-consistent-logging).
```

## 8.3 Monitoring

> Monitoring is only required when running in Microservice mode. As the reconciler is deployed in KCP landscapes, the available Kyma observability stack in KCP will be used.

Reviewer: ```suggestion
Monitoring is only needed when the reconciler runs in microservice mode. Because the reconciler is deployed in KCP landscapes, the available Kyma observability stack in KCP is used.
```

> [Custom metrics](https://github.com/kyma-project/community/tree/main/concepts/observability-custom-metrics) will be exposed to the central Prometheus monitoring system.

Reviewer: ```suggestion
[Custom metrics](https://github.com/kyma-project/community/tree/main/concepts/observability-custom-metrics) are exposed to the central Prometheus monitoring system.
```

# 9. Architecture Decisions

## 9.1 Relational database as central storage

[Postgres](https://www.postgresql.org/) is already used in KCP by other services (KEB and provisioner).

> To avoid adding another storage system, the storage layer of the reconciler will also use Postgres as RDBMS.

Reviewer: ```suggestion
To avoid adding another storage system, the storage layer of the reconciler also uses Postgres as RDBMS.
```

## 9.2 Tasks triggered in intervals

> Modules in the reconicler architecture are decoupled units with particular responsibilities respectively task(s).

Reviewer: ```suggestion
Modules in the reconciler architecture are decoupled units with particular responsibilities or tasks, respectively.
```

> Each module has to perform its task regularly and will, for the beginning, be triggered in intervals (using [Go Tickers](https://gobyexample.com/tickers)).

Reviewer: ```suggestion
Each module must perform its task regularly. In the initial setup, each module is triggered in intervals (using [Go Tickers](https://gobyexample.com/tickers)).
```

> In a later evolution stage, these time based triggers will be replaced by an event-driven approach: a module will be called just in time when it has to perform its task.

Reviewer: ```suggestion
Later, these time-based triggers will be replaced by an event-driven approach: A module will be called just in time when it must perform its task.
```

> ## 9.3 Pre-requisite components

Reviewer: ```suggestion
## 9.3 Prerequisite components
```
https://www.merriam-webster.com/dictionary/prerequisite

In Kyma 1 was agreed that particular components will be installed first (so called `pre-requisites`). A common Kyma component can expect that these pre-requisites are available when it gets deployed. To avoid bigger refactorings of existing Kyma components, the reconciler takes care that these pre-requisites are still supported in Kyma 2. The reconciler installs pre-requisites before common Kyma
components will be deployed.

Pre-requisites are:

* CRDs
* Cluster essential component
* Istio component
* Certificates

> The list of pre-components is defined in the [reconciler configuration](https://github.com/kyma-incubator/reconciler/blob/main/configs/reconciler.yaml#L41-L42).

Reviewer: ```suggestion
The list of prerequisite components is defined in the [reconciler configuration](https://github.com/kyma-incubator/reconciler/blob/main/configs/reconciler.yaml#L41-L42).
```

## 9.2 Tasks triggered in intervals

Modules in the reconciler architecture are decoupled units with particular responsibilities or tasks, respectively.

Each module has to perform its task regularly and will, for the beginning, be triggered in intervals (using [Go Tickers](https://gobyexample.com/tickers)).

In a later evolution stage, these time based triggers will be replaced by an event-driven approach: a module will be called just in time when it has to perform its task.

## 9.3 Pre-requisite components

In Kyma release 1.x, particular components (called "prerequisites") were installed first. Any common Kyma component can expect that these prerequisites are available when it is deployed. 
To avoid major refactoring for Kyma 2.x, the reconciler takes care of this task and installs the prerequisites before common Kyma components are deployed.

Pre-requisites are:

* CRDs
* Component: `cluster essential`
* The Istio component
* Certificates

The list of prerequisite components is defined in the [reconciler configuration](https://github.com/kyma-incubator/reconciler/blob/main/configs/reconciler.yaml#L41-L42).

## 9.4 Software layers

> The code structure of the different reconciler module is following a layered model:

Reviewer: ```suggestion
The code structure of the different reconciler modules follows a layered model:
```

![reconciler configuration](docs/assets/reconciler-swlayers.png).

> Each software layer is allowed to access layers on lower levels. But it is not allowed to access functionalities which are part of a higher layer.

Reviewer: ```suggestion
Each software layer is allowed to access layers on lower levels, but it is not allowed to access functionalities that are part of a higher layer.
```

1. Storages

   Persistent data sinks (e.g. Github, PostgresDB, Cloud storages)

2. Persistency

>    Abstraction layer to access underneath data storages.

Reviewer: ```suggestion
   Abstraction layer to access lower-level data storages.
```

3. Business model

   Including business models.

4. Repository

   Manages business model entities.

5. Business Components

   Business entities offering enhanced business functions and processes.

6. Endpoints

>    Exposed API interfaces (e.g. HTTP endpoints).

Reviewer: ```suggestion
   Exposed API interfaces (such as HTTP endpoints).
```
